matchmaker.js

```diff
@@ -0,0 +1,146 @@
+              "minimum_time_per_choice" : 10}
+        }
+    
+    for (queue_type in this.match_types.keys) {
+            this.match_queues[queue_type] = {
+                "has_player_waiting" : false,
+                "waiting_player" : player = null,
+                "waiting_json_data" : object = null,
+                "waiting_version" : Number
+            }
+        }
+    }
+
+    remove_from_matchmaking(player) {
+        for (queue in this.match_queues.keys) {
+            if (queue["waiting_player"] == player) {
+                this.match_queues[queue]["has_player_waiting"] = false
+            }
+        }
+    }
+
+    join_matchmaking(ws, json_data) {
+        // Check if json_data is an object
+        if (typeof json_data !== 'object' || json_data === null) {
+            console.log("join_matchmaking json is not an object")
+            return false
+        }
+        // Check if 'deck_id' field exists in the object
+        if (!('deck_id' in json_data)) {
+            console.log("join_matchmaking does not have 'deck_id' fields")
+            return false
+        }
+        if (!(typeof json_data.deck_id === 'string' && typeof json_data.deck_id === 'string')) {
+            console.log("join_matchmaking 'deck_id' fields are not strings")
+            return false
+        }
+        if (!('version' in json_data && typeof json_data.version === 'string')) {
+            console.log("join_matchmaking does not have 'version' field")
+            return false
+        }
+        if (!('match_type' in json_data && typeof json_data.match_type === 'string')) {
+            console.log("join_matchmaking does not have 'match_type' field")
+        }
+
+        var joining_player = active_connections.get(ws)
+        if (joining_player === undefined) {
+            console.log("join_matchmaking Player is undefined")
+            return false
+        }
+
+        var player_join_version = json_data.version
+        var deck_id = json_data.deck_id
+        joining_player.set_deck_id(deck_id)
+        var desired_match_type = json_data.match_type
+        
+        if (this.banned_characters.includes(deck_id)) {
+            const message = {
+                type: 'matchmaking_failed',
+                reason: 'banned_matchmaking_character'
+            }
+            ws.send(JSON.stringify(message))
+            return true
+        }
+
+        var success = false
+
+        if (this.match_queues[desired_match_type]["has_player_waiting"] == false) {
+            match_queues[desired_match_type]["has_player_waiting"] = true
+            match_queues[desired_match_type]["waiting_player"] = joining_player
+            match_queues[desired_match_type]["waiting_json_data"] = json_data
+            match_queues[desired_match_type]["waiting_version"] = player_join_version
+            success = true
+        } else {
+            var starting_timer = this.match_types[desired_match_type][starting_timer]
```

Comment:
maybe a comment here that this queue has someone in it, so attempt to start the game